src/units/abilities.cpp

```diff
@@ -402,39 +408,68 @@ bool unit::has_ability_type(const std::string& ability) const
 	return !abilities_.child_range(ability).empty();
 }
 
-
-std::pair<int,map_location> unit_ability_list::highest(const std::string& key, int def) const
+namespace {
+	
+	
+template<typename T, typename TFuncFormula>
+class get_ability_value_visitor : public boost::static_visitor<T>
 {
-	if ( cfgs_.empty() ) {
-		return std::make_pair(def, map_location());
-	}
-	// The returned location is the best non-cumulative one, if any,
-	// the best absolute cumulative one otherwise.
-	map_location best_loc;
-	bool only_cumulative = true;
-	int abs_max = 0;
-	int flat = 0;
-	int stack = 0;
-	for (unit_ability const &p : cfgs_)
+public:
+	// Constructor stores the default value.
+	get_ability_value_visitor(T def, const TFuncFormula& formula_handler) : def_(def), formula_handler_(formula_handler) {}
+
+	T operator()(boost::blank const &) const { return def_; }
+	T operator()(bool)                 const { return def_; }
+	T operator()(int i)                const { return static_cast<T>(i); }
+	T operator()(unsigned long long u) const { return static_cast<T>(u); }
+	T operator()(double d)             const { return static_cast<T>(d); }
+	T operator()(t_string const &)     const { return def_; }
+	T operator()(const std::string& s) const
 	{
-		int value = (*p.first)[key].to_int(def);
-		if ((*p.first)["cumulative"].to_bool()) {
-			stack += value;
-			if (value < 0) value = -value;
-			if (only_cumulative && value >= abs_max) {
-				abs_max = value;
-				best_loc = p.second;
-			}
-		} else if (only_cumulative || value > flat) {
-			only_cumulative = false;
-			flat = value;
-			best_loc = p.second;
+		utils::string_view formula_prefix("formula:");
```

Comment:
> Uh, you really don't need formula/Lua support for add or sub. They're only needed because we don't have formula/Lua support. If you were using formulas or Lua, you'd just use value instead. For example, add=5 becomes value="(base_value + 5)".


That's was actuall my first plan, but it seems like add= sub= stc. are handled in a speicifc way that makes it indepnednt on the order in which the abilities appear, so value="(base_value + 5)". will behave diffrent than add=5 if multiple weapon specialy are invloved. (to abilitites with `add=5` will sum up to add=10 while for two abilitites with both `value="(base_value + 5)"` the game will take the higher oen which is still 5)


> Furthermore, you don't necessarily need a lua_function for every attribute, because Lua functions can return multiple values

This probably requires me to write this code in a case by case (where case is a aspeific ability) basis.

> Please don't use bad syntax just because of concerns of speed

speedwise it wouldn't make a difference between a `formula:` and a `(` prefix. I just